src/Controller/Controller.php

```diff
@@ -757,12 +758,79 @@ public function render(?string $template = null, ?string $layout = null): Respon
         if ($builder->getTemplate() === null) {
             $builder->setTemplate($this->request->getParam('action'));
         }
+        $viewClass = $this->chooseViewClass();
+        $view = $this->createView($viewClass);
 
-        $view = $this->createView();
         $contents = $view->render();
-        $this->setResponse($view->getResponse()->withStringBody($contents));
+        $response = $view->getResponse()->withStringBody($contents);
 
-        return $this->response;
+        $viewContentType = $view->getContentType();
+        $responseType = $response->getHeaderLine('Content-Type');
+        if ($viewContentType && ($responseType === '' || substr($responseType, 0, 9) === 'text/html')) {
+            $response = $response->withHeader('Content-Type', $viewContentType);
+        }
+
+        return $this->setResponse($response)->response;
+    }
+
+    /**
+     * Get the list of View classes this controller can negotiate with.
+     *
+     * Each view class must implement the `getContentType()` hook method
+     * to participate in negotiation.
+     *
+     * @see Cake\Http\ContentTypeNegotiation
+     * @return array<string>
+     */
+    public function getViewClasses(): array
+    {
+        return [];
+    }
+
+    /**
+     * Use the view classes defined on this controller to view
+     * selection based on content-type negotiation.
+     *
+     * TODO: Should this also consider $request->getParam('_ext') as well?
+     *
+     * @return string|null The chosen view class or null for no decision.
+     */
+    protected function chooseViewClass(): ?string
+    {
+        $possibleViewClasses = $this->getViewClasses();
+        if (empty($possibleViewClasses)) {
+            return null;
+        }
+        if ($this->viewBuilder()->getClassName() !== null) {
+            return null;
+        }
+
+        $typeMap = [];
+        foreach ($possibleViewClasses as $class) {
+            $viewContentType = $class::getContentType();
```

Comment:
This means custom view classes won't work with 4.4. without adding (or overriding to return proper value) this method. So that breaks BC.

Reply:
But controller will also have to implement the `viewClasses()` method. My thinking was that when applications were updated to implement `viewClasses()` they could also update their view classes to support `getContentType()`. For plugin views that are lagging behind developers could subclass and add the method as a shim.

RequestHandler features aren't being moved/removed allowing more cautious applications to have full backwards compatibility.

Reply:
Okay.